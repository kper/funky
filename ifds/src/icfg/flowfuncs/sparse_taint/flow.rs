use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
