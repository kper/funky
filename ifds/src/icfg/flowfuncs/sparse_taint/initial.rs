use crate::icfg::tabulation::sparse::defuse::DefUseChain;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintInitialFlowFunction;

impl SparseInitialFlowFunction for SparseTaintInitialFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        init_facts: &Vec<Fact>,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let offset = 0;
        let instructions = &function.instructions;

        let init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        debug!("init fact is {:#?}", init_fact);

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            match instruction {
                Instruction::Const(dest, _)
                | Instruction::Unknown(dest)
                | Instruction::Assign(dest, _)
                | Instruction::Unop(dest, _)
                | Instruction::BinOp(dest, _, _)
                | Instruction::Phi(dest, _, _)
                | Instruction::Conditional(dest, _)
                | Instruction::Load(dest, _, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Call(_callee, _, dests) => {
                    // When the analysis starts at a Call Instruction,
                    // then the function doesn't get called.
                    // But all destinations get automatically tainted.

                    let before = vec![init_fact.clone()];
                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }
                    }
                    for dest in dests {
                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Store(_src, offset, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let mem = ctx
                            .state
                            .add_memory_var(function.name.clone(), offset.clone() as usize);

                        let after_var = defuse
                            .demand_inclusive(ctx, &function, &mem.name, pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let mut applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Return(_dests) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: var.clone(),
                            });
                        }
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    panic!("Block or Jump as first instruction is not supported");
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
