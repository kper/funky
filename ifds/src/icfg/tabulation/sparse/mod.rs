/// Algorithm for sparsification
pub mod tabulation;

/// Implementation of the defuse chain
pub mod defuse;

pub use tabulation::*;