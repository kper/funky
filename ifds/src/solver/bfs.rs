use crate::icfg::graph::{Fact, Graph};
use crate::icfg::state::State;
use crate::solver::*;
use std::collections::VecDeque;

use log::debug;

pub struct Bfs;

impl GraphReachability for Bfs {
    fn all_sinks(&mut self, graph: &mut Graph, state: &State, req: &Request) -> Vec<Taint> {
        let mut queue: VecDeque<&Fact> = VecDeque::new();
        let mut seen = Vec::new();

        if let Some(start) = state
            .get_facts_at(&req.function, req.pc)
            .unwrap()
            .find(|x| &x.belongs_to_var == req.variable.as_ref().unwrap())
        {
            debug!("Adding start {:#?}", start);
            queue.push_back(start);
        } else {
            return Vec::new();
        }

        while let Some(node) = queue.pop_front() {
            debug!("Popping node {:?}", node);
            seen.push(node);
            for child in graph
                .edges
                .iter()
                .filter(|x| x.get_from() == node)
                .map(|x| x.to())
            {
                //debug!("Adding child {:#?}", graph.query_by_fact_id(child));

                if !seen.contains(&child) {
                    debug!("queue child {:?}", child);
                    queue.push_back(child);
                }
            }
        }

        seen.into_iter()
            .map(|x| Taint {
                variable: x.belongs_to_var.clone(),
                pc: x.next_pc,
                function: x.function.clone(),
            })
            .collect()
    }
}
