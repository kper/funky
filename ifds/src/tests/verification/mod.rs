use itertools::Itertools;

use log::debug;

use crate::grammar::*;
use crate::icfg::tabulation::fast::TabulationFast;
use crate::icfg::tabulation::naive::TabulationNaive;
use crate::icfg::tabulation::orig::TabulationOriginal;
use crate::icfg::tabulation::sparse::TabulationSparse;

use std::collections::HashSet;

use crate::icfg::flowfuncs::sparse_taint::flow::SparseTaintNormalFlowFunction;
use crate::icfg::flowfuncs::sparse_taint::initial::SparseTaintInitialFlowFunction;
use crate::icfg::flowfuncs::taint::flow::TaintNormalFlowFunction;
use crate::icfg::flowfuncs::taint::initial::TaintInitialFlowFunction;
use crate::icfg::tikz::render_to;
use crate::icfg::tikz2::render_to as render_to2;
use insta::assert_snapshot;
use pretty_assertions::assert_eq;

use crate::solver::bfs::*;
use crate::solver::*;

macro_rules! ir {
    ($ir:expr) => {{
        let prog = ProgramParser::new().parse(&$ir).unwrap();
        prog
    }};
}

macro_rules! naive {
    ($name:expr, $prog:expr, $req:expr) => {{
        let mut convert = TabulationNaive::default();

        let (mut graph, state, _) = convert.visit(&$prog).unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("_naive_{}_dot", $name), output);

        let mut solver = Bfs;
        let mut sinks = solver
            .all_sinks(&mut graph, &state, &$req)
            .into_iter()
            .filter(|x| x.pc > $req.pc + 1)
            .map(|x| x.variable)
            .unique()
            .collect::<Vec<_>>();

        sinks.push("taut".to_string());

        sinks
    }};
}

macro_rules! orig {
    ($name:expr, $prog:expr, $req:expr) => {{
        let mut convert = TabulationOriginal::default();

        let (graph, state) = convert.visit(&$prog, $req).unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("_orig_{}_dot", $name), output);

        let sinks = graph
            .edges
            .iter()
            .map(|x| x.to())
            .filter(|x| &x.function == &$req.function)
            .map(|x| &x.belongs_to_var)
            .unique()
            .map(|x| x.clone())
            .collect::<Vec<_>>();

        sinks
    }};
}

macro_rules! fast {
    ($name:expr, $prog:expr, $req:expr) => {{
        let mut convert = TabulationFast::new(TaintInitialFlowFunction, TaintNormalFlowFunction);

        let (graph, state) = convert.visit(&$prog, $req).unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("_fast_{}_dot", $name), output);

        let sinks = graph
            .edges
            .iter()
            .map(|x| x.to())
            .filter(|x| &x.function == &$req.function)
            .map(|x| &x.belongs_to_var)
            .unique()
            .map(|x| x.clone())
            .collect::<Vec<_>>();

        sinks
    }};
}

macro_rules! sparse {
    ($name:expr, $prog:expr, $req:expr) => {{
        let mut convert = TabulationSparse::new(
            SparseTaintInitialFlowFunction,
            SparseTaintNormalFlowFunction,
        );

        let (graph, state) = convert.visit(&$prog, $req).unwrap();

        let output = render_to2(&graph, &state);

        assert_snapshot!(format!("_sparse_{}_dot", $name), output);

        let sinks = graph
            .edges
            .iter()
            .map(|x| x.to())
            .filter(|x| &x.function == &$req.function)
            .map(|x| &x.belongs_to_var)
            .unique()
            .map(|x| x.clone())
            .collect::<Vec<_>>();

        sinks
    }};
}

macro_rules! run {
    ($name:expr, $ir:expr, $req:expr) => {
        let prog = ir!($ir);
        let mut s1 = naive!($name, &prog, $req);
        s1.sort();
        let mut s2 = orig!($name, &prog, $req);
        s2.sort();
        let mut s3 = fast!($name, &prog, $req);
        s3.sort();
        let mut s4 = sparse!($name, &prog, $req);
        s4.sort();

        debug!("naive {:#?}", s1);
        debug!("orig {:#?}", s2);
        debug!("fast {:#?}", s3);
        debug!("sparse {:#?}", s4);

        let s1 = s1.into_iter().collect::<HashSet<_>>();
        let s2 = s2.into_iter().collect::<HashSet<_>>();
        let s3 = s3.into_iter().collect::<HashSet<_>>();
        let s4 = s4.into_iter().collect::<HashSet<_>>();

        assert!(s1.is_superset(&s2), "naive and orig failed");
        assert!(s1.is_superset(&s3), "naive and fast failed");
        assert!(s1.is_superset(&s4), "naive and sparsed failed");

        assert!(s2.is_superset(&s3), "orig and fast failed");
        assert!(s2.is_superset(&s4), "orig and sparse failed");

        assert_eq!(s1, s2, "naive and orig failed");
        assert_eq!(s1, s3, "naive and fast failed");
        assert_eq!(s1, s4, "naive and sparse failed");
    };
}

#[test]
fn test_intra() {
    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let ir = "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = %0
         };";

    run!("intra_reach", ir, &req);
}

#[test]
fn test_if_else() {
    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let ir = "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = %0
            %2 = 3
            GOTO 3
            BLOCK 2
            %1 = 1
            GOTO 3
            BLOCK 3
            %3 = %1 op %0   
        };";

    run!("verification_if_else", ir, &req);
}

#[test]
fn test_early_return() {
    let req = Request {
        function: "test".to_string(),
        pc: 0,
        variable: Some("%0".to_string()),
    };

    let ir = "\
        define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            %1 <- CALL mytest(%0)
            %2 <- CALL mytestfoo(%0)
            %3 <- CALL mytestfoo(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1) {
            %0 = 2   
            %1 = 3
            RETURN %1;
        };
        define mytestfoo (param %0) (result 1) (define %0 %1) {
            %1 = 3
            RETURN %0;
        };
    ";

    run!("verification_early_return", ir, &req);
}

#[test]
fn test_memory_load_different_functions2() {
    let req = Request {
        variable: Some("%1".to_string()),
        function: "0".to_string(),
        pc: 2,
    };
    let ir = "
       define 0 (result 0) (define %0 %1 %2 %3 %4 %5 %6 %7) {
        BLOCK 0
        %0 = 8
        %1 = -12345
        STORE FROM %1 OFFSET 0 + %0 ALIGN 2 32
        %2 <- CALL 1 ()
        STORE FROM %1 OFFSET 1 + %0 ALIGN 2 32
        %3 <- CALL 1 ()
        RETURN ;
       }; 

       define 1 (result 1) (define %0 %1) {
        %1 = 8
        %0 = LOAD OFFSET 0 + %1 ALIGN 0
        RETURN %0;
       };
    ";
    run!("memory_load_different_functions2", ir, &req);
}

#[test]
fn test_ir_return_double() {
    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let ir = "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 %2 <- CALL mytest(%0)
            %1 = 2
        };
        define mytest (param %0) (result 2) (define %0 %1) {
            %0 = 2
            %1 = 3
            RETURN %0 %1;
        };
        ";
    run!("return_double", ir, &req);
}

#[test]
fn test_ir_return_double2() {
    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let ir = "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 %2 <- CALL mytest(%0)
            %1 = 2
        };
        define mytest (param %0) (result 2) (define %0 %1) {
            %1 = 3
            RETURN %0 %1;
        };
        ";
    run!("return_double2", ir, &req);
}

#[test]
fn test_global_get_and_set_multiple_functions() {
    let req = Request {
        variable: Some("%1".to_string()),
        function: "0".to_string(),
        pc: 0,
    };
    let ir = "
        define 0 (param %0) (result 0) (define %-2 %-1 %0 %1 %2) {
        %1 = %-1
        %-2 = %1
        %2 = 1
        %0 <- CALL 1 (%2)
        %1 <- CALL 2 ()
        };

        define 1 (param %0) (result 1) (define %-2 %-1 %0) {
        %0 = %-2
        %-1 = 1
        RETURN %0;
        };

        define 2 (result 1) (define %-2 %-1 %0) {
        %0 = 1
        %-2 = 1
        RETURN %0;
        };
    ";
    run!("global_get_and_set_multiple_functions", ir, &req);
}

#[test]
fn test_return_branches() {
    let req = Request {
        variable: Some("%0".to_string()),
        function: "test".to_string(),
        pc: 0,
    };
    let ir = "define test (result 0) (define %0 %1) {
            %0 = 5
            %1 <- CALL mytest(%0)
        };
        define mytest (param %0) (result 1) (define %0 %1 %2) {
            %1 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2
            BLOCK 1
            %1 = 2
            %2 = 3
            RETURN %1;
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            RETURN %0;
        };
        ";

    run!("return_branches", ir, &req);
}
