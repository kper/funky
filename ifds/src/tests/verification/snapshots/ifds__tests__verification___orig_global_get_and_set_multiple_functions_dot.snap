---
source: src/tests/verification/mod.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (27) at (1, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (28) at (1, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (29) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (30) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (31) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (32) at (0, 3) { };
\node[font=\tiny] at (-1.5, 0.5) {Const(0, 1.0)};
\node[font=\tiny] at (-1.5, 1.5) {Const(-2, 1.0)};
\node[font=\tiny] at (-1.5, 2.5) {Return([0])};
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (15) at (7, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (16) at (7, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (17) at (7, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (18) at (7, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (19) at (7, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (20) at (6, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (21) at (6, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (22) at (6, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (23) at (5, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (24) at (5, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (25) at (5, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (26) at (5, 3) { };
\node[font=\tiny] at (3.5, 0.5) {Assign(0, -2)};
\node[font=\tiny] at (3.5, 1.5) {Const(-1, 1.0)};
\node[font=\tiny] at (3.5, 2.5) {Return([0])};
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (0) at (13, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (1) at (13, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%-2$}] (2) at (13, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (3) at (12, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%0$}] (4) at (12, 5) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (5) at (11, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (6) at (11, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (7) at (11, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $\%1$}] (8) at (11, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (9) at (10, 0) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (10) at (10, 1) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (11) at (10, 2) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (12) at (10, 3) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (13) at (10, 4) { };
\node[circle,fill,inner sep=1pt,label=left:{\scriptsize $taut$}] (14) at (10, 5) { };
\node[font=\tiny] at (8.5, 0.5) {Assign(1, -1)};
\node[font=\tiny] at (8.5, 1.5) {Assign(-2, 1)};
\node[font=\tiny] at (8.5, 2.5) {Const(2, 1.0)};
\node[font=\tiny] at (8.5, 3.5) {Call(1, [2], [0])};
\node[font=\tiny] at (8.5, 4.5) {Call(2, [], [1])};
		\path[->, blue] (15) [loop right] edge  node { } (15);
		\path[->, blue] (15) [bend right] edge  node { } (16);
		\path[->, blue] (15) [bend right] edge  node { } (18);
		\path[->, blue] (15) [bend right] edge  node { } (17);
		\path[->, blue] (15) [bend right] edge  node { } (19);
		\path[->, blue] (15) [bend right] edge  node { } (20);
		\path[->, blue] (15) [bend right] edge  node { } (21);
		\path[->, blue] (15) [bend right] edge  node { } (22);
		\path[->, blue] (27) [loop right] edge  node { } (27);
		\path[->, blue] (27) [bend right] edge  node { } (28);
		\path[->, blue] (9) [bend right] edge  node { } (0);
		\path[->, blue] (9) [bend right] edge  node { } (1);
		\path[->, blue] (9) [bend right] edge  node { } (2);
		\path[->, blue] (9) [bend right] edge  node { } (3);
		\path[->, blue] (9) [bend right] edge  node { } (4);
		\path[->, blue] (9) [bend right] edge  node { } (5);
		\path[->, blue] (9) [bend right] edge  node { } (6);
		\path[->, blue] (9) [bend right] edge  node { } (7);
		\path[->, blue] (9) [bend right] edge  node { } (8);
		\path[->, blue] (9) [loop right] edge  node { } (9);
		\path[->, blue] (9) [bend right] edge  node { } (10);
		\path[->, blue] (9) [bend right] edge  node { } (11);
		\path[->, blue] (9) [bend right] edge  node { } (12);
		\path[->, blue] (9) [bend right] edge  node { } (13);
		\path[->, blue] (9) [bend right] edge  node { } (14);
		\path[->, blue] (23) [loop right] edge  node { } (23);
		\path[->, blue] (23) [bend right] edge  node { } (24);
		\path[->, blue] (23) [bend right] edge  node { } (25);
		\path[->, blue] (23) [bend right] edge  node { } (26);
		\path[->, blue] (29) [loop right] edge  node { } (29);
		\path[->, blue] (29) [bend right] edge  node { } (30);
		\path[->, blue] (29) [bend right] edge  node { } (31);
		\path[->, blue] (29) [bend right] edge  node { } (32);


                \end{tikzpicture}
            \end{document}
